spec/bolt/last_failure_spec.rb

```diff
@@ -0,0 +1,209 @@
+    end
+
+    it 'runs a command with success targets' do
+      expect(executor).to receive(:run_command)
+        .with([targets[0]], 'whoami', kind_of(Hash))
+        .and_return(result_set)
+      run_cli(['command', 'run', 'whoami', '--retry', 'success'])
+    end
+
+    it 'fails with an unhandled filter' do
+      expect do
+        run_cli(['command', 'run', 'whoami', '--retry', 'invalid'])
+      end.to raise_error(/Unexpected option/)
+    end
+  end
+
+  context 'with an empty last_failure file' do
+    before(:each) do
+      write_last_failure(['original result'])
+
+      allow(executor).to receive(:start_plan)
+      allow(executor).to receive(:finish_plan)
+      allow(pal).to receive(:parse_params)
+      allow(pal).to receive(:parse_manifest)
+    end
+
+    let(:success_set) { Bolt::ResultSet.new([result_set.results[0], result_set.results[0]]) }
+
+    it 'updates the file when a command fails' do
+      allow(executor).to receive(:run_command)
+        .with(targets, 'whoami', kind_of(Hash))
+        .and_return(result_set)
+      run_cli(['command', 'run', 'whoami', '--nodes', target_spec.join(',')])
+
+      expect(read_last_failure).to eq(failure_array)
+    end
+
+    it 'does not update the file with --no-save-failures' do
+      allow(executor).to receive(:run_command)
+        .with(targets, 'whoami', kind_of(Hash))
+        .and_return(result_set)
+      run_cli(['command', 'run', 'whoami', '--no-tty', '--no-save-failures', '--nodes', target_spec.join(',')])
+
+      expect(read_last_failure).to eq(['original result'])
+    end
+
+    it 'does not update the file with save-failures: false' do
+      File.write(@boltdir.config_file, { 'save-failures' => false }.to_yaml)
+
+      allow(executor).to receive(:run_command)
+        .with(targets, 'whoami', kind_of(Hash))
+        .and_return(result_set)
+      run_cli(['command', 'run', 'whoami', '--no-save-failures', '--nodes', target_spec.join(',')])
+
+      expect(read_last_failure).to eq(['original result'])
+    end
+
+    it 'does not update the file when a command succeeds' do
+      allow(executor).to receive(:run_command)
+        .with(targets, 'whoami', kind_of(Hash))
+        .and_return(success_set)
+      run_cli(['command', 'run', 'whoami', '--nodes', target_spec.join(',')])
+
+      expect(read_last_failure).to eq(['original result'])
+    end
+
+    it 'updates the file when a plan fails with a ResultSet' do
+      expect(pal).to receive(:run_plan).and_return(result_set)
+      run_cli(%w[plan run whoami])
+
+      expect(read_last_failure).to eq(failure_array)
+    end
+
+    it 'does not update the file when a plan fails with nil' do
+      expect(pal).to receive(:run_plan)
+        .and_return(Bolt::PlanResult.new('oops', 'failure'))
+      run_cli(%w[plan run whoami])
+
+      expect(read_last_failure).to eq(['original result'])
+    end
+
+    it 'does not update the file when a plan fails with nil' do
```

Comment:
Copy paste error? This looks more like `does not update the file when a plan succeeds`.